File: Problem3.c

#include <stdio.h>
#include <math.h>

#define NUMBER 600851475143

int isPrime(int test) {
    int calculateTo = ( (int) floor(sqrt(test)) );

    if(test == 1) {
        return 0; 
    }

    if(test < 4) {
        return 1; 
    }

    if(test % 2 == 0) {
        return 0; 
    }

    if(test < 9) {
        return 1; 
    }

    if(test % 3 == 0) {
        return 0; 
    }

    for(int i = 5; i <= calculateTo; i += 2) {
        if(test % i == 0) {
            return 0; 
        }
        if(test % (i + 2) == 0) {
            return 0; 
        }
    }

    return 1;
}

int main() {
    int highest = 0;

    for(int i = 3; i <= floor(sqrt(NUMBER)); i += 2) {
        if(isPrime(i)) {
            if(NUMBER % i == 0)
                highest = i;
        }

    }

    printf("%d", highest);
    
    return 0;

}

File: Problem3.go

package main

import "math"
import "fmt"

func is_prime(test int) bool {
    calculate_to := int(math.Sqrt(float64(test)))

    if test == 1 {
        return false 
    }

    if test < 4 {
        return true 
    }

    if test % 2 == 0 {
        return false 
    }

    if test < 9 {
        return true 
    }

    if test % 3 == 0 {
        return false 
    }

    x := 5

    for x < calculate_to {
        if test % x == 0 {
            return false 
        }
        if test % (x + 2) == 0 {
            return false 
        }

        x = x + 6
    }

    return true
}

func main() {
    number := 600851475143
    end := int(math.Sqrt(float64(number)))
    highest := 0
    x := 3

    for x < end {
        if is_prime(x) {
            if number % x == 0 {
                highest = x 
            } 
        } 

        x = x + 2
    }

    fmt.Println("Highest", highest)

}
